src/sc/sidechain.cpp

```diff
@@ -783,108 +782,91 @@ void CSidechain::UpdateScFees(const CScCertificateView& certView, int blockHeigh
-                scFees_v2.pop_front();
+                const auto& entry = scFees.front();
+                LogPrint("sc", "%s():%d - popping %s from list, as scFees maxsize has been reached\n",
+                    __func__, __LINE__, entry->ToString().c_str());
+                scFees.pop_front();
             }
 
             // Also purge all elements submitted within a block that now is too old to be considered
             const auto threshold{blockHeight - this->maxSizeOfScFeesContainers};
-            scFees_v2.remove_if([threshold](Sidechain::ScFeeData_v2 entry) {
-                const bool testCondition = (entry.submissionHeight <= threshold);
+            scFees.remove_if([threshold](std::shared_ptr<Sidechain::ScFeeData> entry) {
+                const std::shared_ptr<Sidechain::ScFeeData_v2> casted_entry = std::dynamic_pointer_cast<Sidechain::ScFeeData_v2>(entry);
+                const bool testCondition = (casted_entry->submissionHeight <= threshold);
                 if (testCondition)
-                    LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as entry is too old (threshold height was %d)\n",
-                        __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight, threshold);
+                    LogPrint("sc", "%s():%d - popping %s from list, as entry is too old (threshold height was %d)\n",
+                        __func__, __LINE__, casted_entry->ToString().c_str(), threshold);
                 return testCondition;
             });
+
         }
         // On the other hand, if we found a scFeeData element for the current height, we just update it with lower
         // ft and/or mbtr fee rates. All the checks on the container size and element age have been performed
         // at the moment of the element insertion, hence their are not required here
         else {
-            scFeeIt->forwardTxScFee = std::min(scFeeIt->forwardTxScFee, ftScFee);
-            scFeeIt->mbtrTxScFee = std::min(scFeeIt->mbtrTxScFee, mbtrScFee);
+            scFeeIt->get()->forwardTxScFee = std::min(scFeeIt->get()->forwardTxScFee, ftScFee);
+            scFeeIt->get()->mbtrTxScFee    = std::min(scFeeIt->get()->mbtrTxScFee, mbtrScFee);
         }
+
     }
 }
 
 void CSidechain::DumpScFees() const
 {
-    if (!isNonCeasing()) {
-        for (const auto& entry : scFees) {
-            std::cout << "[" << std::setw(2) << entry.forwardTxScFee
-                      << "/" << std::setw(2) << entry.mbtrTxScFee << "]";
-        }
-    }
-    else {
-        for (const auto& entry : scFees_v2) {
-            std::cout << "[" << std::setw(2) << entry.forwardTxScFee
-                      << "/" << std::setw(2) << entry.mbtrTxScFee
-                      << "/" << std::setw(2) << entry.submissionHeight << "]";
-        }
+
+    for (const auto& entry : scFees) {
+        std::cout << entry->ToString() << std::endl;
     }
-    std::cout << std::endl;
+
```

Comment:
This is actually very expensive as the stream is flushed at every iteration.
Use this instead
```
    for (const auto& entry : scFees) {
        std::cout << entry->ToString() << "\n";
    }
    std::cout << std::flush;
```
```

Reply:
Given that this function is definitely not on performance critical path (it would not make sense to print to `cout` otherwise), and it's used only in gtests where it might be better to flush immediately for sake of debugging, I'd rather leave it as it is.

Reply:
This comment doesn't make any sense to me.

Reply:
I agree with @drgora. `DumpScFees()` is used only in two tests, while any other dump of the `scFees` list is done  through `LogPrint()`. I don't think that performance is an issue here, hence I propose to keep the function as it is.

Reply:
@AndreaLanfranchi sorry I was not really clear.
What I meant is that given that this function is not, and will never be in its current form, on any performance critical path, I'd rather keep it more compact. I agree it's not a big difference in size, as surely it won't be a big difference in performance anyway.
About debugging, I was suggesting that not delaying debugging logs might help understand the actual sequence of events, especially in multithreaded applications. Again, not a strong point, I agree, but also no strong points against the current form.